usb/backend/libusb0.py

```diff
@@ -619,27 +619,47 @@ def reset_device(self, dev_handle):
 
     @methodtrace(_logger)
     def is_kernel_driver_active(self, dev_handle, intf):
+        # libusb0's usb_get_driver_np() is platform specific / not portable.
+        # This function checks what kernel module handles dev_handle/intf.
+        # Various implementations may return system specific values - for
+        # instance Linux may return errno.ENODATA not present on FreeBSD.
+        # On FreeBSD successul call returns 0 otherwise -1.
+
         if not hasattr(_lib, 'usb_get_driver_np'):
             raise NotImplementedError(self.is_kernel_driver_active.__name__)
-        from errno import ENODATA
         buf = usb.util.create_buffer(_USBFS_MAXDRIVERNAME + 1)
         name, length = buf.buffer_info()
         length *= buf.itemsize
-        # based on the implementation of libusb_kernel_driver_active
+
+        # Linux: based on the implementation of libusb_kernel_driver_active
         # (see libusb/os/linux_usbfs.c @@ op_kernel_driver_active):
         # usb_get_driver_np fails with ENODATA when no kernel driver is bound,
         # and if 'usbfs' is bound that means that a userspace program is
         # controlling the device (e.g. using this very library)
+        if "linux" in sys.platform:
+            try:
+                from errno import ENODATA
+            except:
+                raise NotImplementedError(self.is_kernel_driver_active.__name__)
+
         try:
             _check(_lib.usb_get_driver_np(
                         dev_handle,
                         intf,
                         cast(name, c_char_p),
                         length))
-            return cast(name, c_char_p).value != b'usbfs'
+            if "linux" in sys.platform:
+                # On Linux 'usbfs' indicates userland application.
+                return cast(name, c_char_p).value != b'usbfs'
+            else:
+                # On other platforms just check any module attached.
+                return buf.index(0) != 0
```

Comment:
Have you confirmed that this behaves as `libusb_kernel_driver_active()`?

> Determine if a kernel driver is active on an interface. If a kernel driver is active, you cannot claim the interface, and libusb will be unable to perform I/O.

Specifically, `is_kernel_driver_active()` should return `False` if the kernel driver bound to the device is the one this backend uses (or will use anyway).

This is counter intuitive given the function name, but the idea is that this method, like the equivalent libusb-1 API, is used to check whether something has to be done about the current kernel driver before being able to interact with the device.

Reply:
* Cannot say for every OS but on FreeBSD we can just write do /dev/usb*` from the user standpoint if file permission allows it. 
* User should not be allowed to load/unload kernel modules, only root can do that, so unloading kernel module will always fail when performed by user application. 
* This is why in `pyOCD` failed unloading module will drop a warning instead application failure. Then it tried to write to a device anyway and it works just fine.
* On FreeBSD `uhid` kernel module is attached to the `CMSIS-DAP` interface.

From what I can see only FreeBSD is using its own LibUSB implementation that "joins" 0.1 and 1.0 functionalities. OpenBSD has `libusb-1.0` package and has no `ENODATA` defined. NetBSD has `ENODATA` defined and `libusb-0.1.12nb7`, `libusb-compat-0.1.7`, `libusb1-1.0.24`, as well as `py38-libusb1-1.7.1` and `libusb-0.1.12nb7` packages. But I guess their behavior is similar - user can just write to `/dev/usb*` [1][2][3]. I only have VM with OpenBSD and NetBSD if necessary I can verify these functions on real hardware..?

Is this correct @hselasky ? :-)

I have tried to get a common denominator here :-)

[1] https://nixdoc.net/man-pages/FreeBSD/man4/usb.4.html
[2] https://man.openbsd.org/usb.4
[3] https://man.netbsd.org/usb.4

Reply:
OpenBSD also has libusb-1.0 and libusb-compat.
http://cvsweb.openbsd.org/cgi-bin/cvsweb/ports/devel/libusb1/
http://cvsweb.openbsd.org/cgi-bin/cvsweb/ports/devel/libusb-compat/

The libusb-0.1 is obsolete in OpenBSD.
http://cvsweb.openbsd.org/cgi-bin/cvsweb/ports/devel/libusb/

Reply:
NetBSD have libusb-1.0.
https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/devel/libusb1/

NetBSD keeps both libusb-compat-0.1 and legacy libusb-0.1
https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/devel/libusb-compat/
https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/devel/libusb/